index.rst

```diff
@@ -50,6 +50,133 @@
+  This is fragile but hard to fix, unless we just prohibit that kind of ``RUN`` collection extension or do enumerate all relevant software somehow up front.
+
+- The approach of using datasets to record arbitrary provenance for higher-level tooling is currently extensible only to the extent that those datasets can use existing storage classes.
+  New storage classes must currently be registered by editing configuration in the data repository or ``daf_butler`` manually.
+  This is a problem for much more than just provenance, however (it also limits pipeline extensibility), and is something we hope to resolve.
+
+- Using ``CHAINED`` collections to link input collections to output collections is often assumed to relate input and output dataset the share the same data ID (or have related data IDs).
+  This *usually* works, at least as far as relating predicted input to predicted output datasets, but only because our ``QuantumGraph`` generation algorithm is based on data ID commonality; some other ``QuantumGraph`` generation algorithm that is equally valid for execution may not, or may have important exceptions.
+  It also fails when a ``CHAINED`` collection includes a "merge" of two output ``RUN`` collections that have conflicting datasets (i.e. they have same dataset type and data ID); dataset lookup then uses the order in which those ``RUN`` collections appear in the ``CHAINED`` collection to resolve the conflict, but this may not correspond to what was actually used as an input, and in fact more than one of the conflictig datasets may have been used as inputs (to different quanta).
+
+Saving complete quantum provenance
+==================================
+
+Representing executed quanta in Registry
+----------------------------------------
+
+TODO.
+
+Sketch the schema, describe predicted/available/actual input and predicted/actual output distinctions.
+Plan to use UUIDs, even for predicted-only datasets.
+Need to figure out how to work empty-data-ID datasets in.
+
+Recording provenance during execution
+-------------------------------------
+
+TODO:
+
+- Write predicted QG to files or Redis then write new per-quantum output files or extend info in Redis; consider this storage to be in repo, but in a third sibling to Registry and Datastore rather than either.
+- Bring home provenance to Registry when job completes and datasets are also brought home.
+- Defer most of this description to another technote.
+
+Interfaces for querying quantum provenance
+------------------------------------------
+
+TODO:
+
+- Support getting QG given (all optional): input DatasetRef/DatasetType/Task/Quantum, output DatasetRef/DatasetType/Task/Quantum, collections all datasets or quanta must be in.
+- Need to think about whether QG is well-defined for all argument combinations, and what arguments really mean; "input" and "outputs" are only well-defined for linear graphs, but there are probably definitions with that limiting behavior we can assume.
```

Comment:
We should consider starting to use provenance nomenclature for provenance (as used by IVOA provenance model which is derived from the W3C provenance model) to be explicit about our meaning. (https://www.ivoa.net/documents/ProvenanceDM/20200411/REC-ProvenanceDM-1.0-20200411.pdf)

Reply:
I think I picked up the "predicted" vs "actual" terminology from @MichelleGower , who I think got it from some (other? same?) standardized set of provenance definitions.  I did make up "available" on my own, though.  I'll scan that link to see if anything equivalent jumps out, unless Michelle can save me some time by clarifying where "predicted" and "actual" come from.

Reply:
I would have said @TallJimbo came up with the names.  I definitely remember talking about the difference between the two categories and whether LSST wanted to track both, but not the origins of the names.

Reply:
The IVOA provenance model seems to map to our `QuantumGraph` as follows:

- a `quantum` is one kind of (here, the only relevant kind of) _Activity_;
- a `dataset` is one kind of (here, the only relevant kind of) _Entity_;
- a `quantum` _Activity_ _Used_ an `input` `dataset`;
- an `output` `dataset` _Entity_ _WasGeneratefBy a `quantum` _Activity_.

IVOA also has:
- a _WasDerivedFrom_ relationship between two _Entities_;
- a _WasInformedBy_ relationship between two _Activities_.

These could be useful in exporting "collapsed" versions of the QG in which we drop datasets or quanta and instead define edges between the one we didn't drop.

We currently have no concept that maps to IVOA's _agent_.  IVOA has no concepts that map to the `predicted`, `actual`, and `available` distinctions I propose here, but there are free-form _UsageDescription_ and _GeneratedDescription_ fields where we can put them.

So, overall, I don't see a clear place to start using IVOA terminology in our own systems, at least not yet; the terms we might have borrowed already have well-established names on our side that are either more specific (`quantum` vs _Activity_ or `dataset` vs. _Entity_) or have an adjective form that fits better where we use the term (`input` or `output`) than a past-tense verb (_Used_, _WasGeneratedBy_).

I'll include this in the text, and I think it absolutely does make sense to reference the IVOA terms in other documentation, including code documentation.

Reply:
We will have to be able to export provenance in the W3C naming style so we should at least be cognizant of the WasGeneratedBy, WasDerivedFrom scheme.